File: packages/core/src/modules/AgentBuffs.ts
import { HTTPClient } from '../client/http-client'
import { SimpleEventEmitter } from '../utils/event-emitter'

export interface Buff {
  buff_id: string
  name: string
  type: string
  category: string
  is_active: boolean
  expires_at: string
  priority: number
  effects: Record<string, any> // Произвольные пути в формате "data.path.to.field"
  scope?: 'ecosystem' | 'project' | 'user_in_project'
}

export interface ActiveBuffsResponse {
  user_id?: number
  project_id?: number
  active_buffs: Buff[]
}

export interface BuffStatusResponse {
  user_id?: number
  project_id?: number
  has_active_buffs: boolean
  active_buffs: Buff[]
}

export interface EffectiveLimitsResponse {
  user_id?: number
  project_id?: number
  effective_limits: {
    limits: Record<string, number>
    features: Record<string, boolean>
  }
  contributing_buffs?: Array<{
    buff_id: string
    name: string
    scope: string
    priority: number
  }>
}

export interface CreateBuffRequest {
  entity_id: number
  entity_kind: 'user' | 'project'
  name: string
  type: string
  category: string
  duration_days: number
  priority?: number
  effects: Record<string, any> // Произвольные пути в формате "data.path.to.field"
  on_apply?: Array<{
    type: 'event' | 'command' | 'processor'
    [key: string]: any
  }>
  on_expire?: Array<{
    type: 'event' | 'command' | 'processor'
    [key: string]: any
  }>
  config?: {
    persistent?: boolean
    revert_on_expire?: boolean
    extends_on_reapply?: boolean
  }
  project_id?: number
}

export interface CreateBuffResponse {
  buff_id: string
  state: string
  created_at: string
}

export interface ApplyBuffResponse {
  buff_id: string
  state: string
  applied_at: string
  expires_at: string
}

export interface ExtendBuffResponse {
  buff_id: string
  state: string
  expires_at: string
}

export interface RevertBuffResponse {
  reverted: boolean
  buff_id: string
  deleted: boolean
}

export interface CancelBuffResponse {
  cancelled: boolean
  buff_id: string
  deleted: boolean
}

export interface RetryBuffResponse {
  buff_id: string
  state: string
  applied_at?: string
  expires_at?: string
  reverted?: boolean
  deleted?: boolean
}

export class AgentBuffs {
  private httpClient: HTTPClient
  private eventEmitter: SimpleEventEmitter

  constructor(httpClient: HTTPClient, eventEmitter: SimpleEventEmitter) {
    this.httpClient = httpClient
    this.eventEmitter = eventEmitter
  }

  /**
   * Получить список активных баффов
   */
  async getActiveBuffs(params: {
    entityId: number
    entityKind: 'user' | 'project'
    category?: string
  }): Promise<ActiveBuffsResponse> {
    const response = await this.httpClient.get<ActiveBuffsResponse>(
      `/buffs/${params.entityKind}/${params.entityId}/active`,
      params.category ? { category: params.category } : undefined
    )
    return response.data
  }

  /**
   * Получить статус баффов
   */
  async getBuffStatus(params: {
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<BuffStatusResponse> {
    const response = await this.httpClient.get<BuffStatusResponse>(
      `/buffs/${params.entityKind}/${params.entityId}/status`
    )
    return response.data
  }

  /**
   * Получить информацию о конкретном баффе
   */
  async getBuff(params: {
    buffId: string
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<Buff> {
    const response = await this.httpClient.get<Buff>(
      `/buffs/${params.buffId}`,
      {
        entity_id: params.entityId,
        entity_kind: params.entityKind
      }
    )
    return response.data
  }

  /**
   * Получить все баффы проекта (включая шаблоны)
   */
  async getProjectAllBuffs(params: {
    projectId: number
    includeEcosystem?: boolean
    category?: string
  }): Promise<{
    project_id: number
    all_buffs: Array<Buff & { state?: string; created_at?: string; applied_at?: string }>
    templates: Array<Buff & { state?: string; created_at?: string }>
    active_buffs: Buff[]
  }> {
    const queryParams: Record<string, any> = {
      include_ecosystem: params.includeEcosystem ?? true
    }
    if (params.category) {
      queryParams.category = params.category
    }
    const response = await this.httpClient.get<{
      project_id: number
      all_buffs: Array<Buff & { state?: string; created_at?: string; applied_at?: string }>
      templates: Array<Buff & { state?: string; created_at?: string }>
      active_buffs: Buff[]
    }>(
      `/buffs/project/${params.projectId}/all`,
      queryParams
    )
    return response.data
  }

  /**
   * Получить активные баффы проекта
   */
  async getProjectActiveBuffs(params: {
    projectId: number
    includeEcosystem?: boolean
    category?: string
  }): Promise<ActiveBuffsResponse> {
    const queryParams: Record<string, any> = {
      include_ecosystem: params.includeEcosystem ?? true
    }
    if (params.category) {
      queryParams.category = params.category
    }
    const response = await this.httpClient.get<ActiveBuffsResponse>(
      `/buffs/project/${params.projectId}/active`,
      queryParams
    )
    return response.data
  }

  /**
   * Получить активные баффы пользователя в проекте
   */
  async getUserInProjectActiveBuffs(params: {
    projectId: number
    userId: number
    includeProjectBuffs?: boolean
    includeEcosystem?: boolean
    category?: string
  }): Promise<ActiveBuffsResponse> {
    const queryParams: Record<string, any> = {
      include_project_buffs: params.includeProjectBuffs ?? true,
      include_ecosystem: params.includeEcosystem ?? false
    }
    if (params.category) {
      queryParams.category = params.category
    }
    const response = await this.httpClient.get<ActiveBuffsResponse>(
      `/buffs/project/${params.projectId}/user/${params.userId}/active`,
      queryParams
    )
    return response.data
  }

  /**
   * Создать новый бафф
   */
  async createBuff(data: CreateBuffRequest): Promise<CreateBuffResponse> {
    const response = await this.httpClient.post<CreateBuffResponse>('/buffs', data)
    return response.data
  }

  /**
   * Обновить существующий бафф (только для шаблонов в состоянии PENDING)
   */
  async updateBuff(buffId: string, data: CreateBuffRequest): Promise<CreateBuffResponse> {
    const response = await this.httpClient.put<CreateBuffResponse>(`/buffs/${buffId}`, data)
    return response.data
  }

  /**
   * Применить бафф
   */
  async applyBuff(params: {
    buffId: string
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<ApplyBuffResponse> {
    const response = await this.httpClient.post<ApplyBuffResponse>(
      `/buffs/${params.buffId}/apply`,
      {
        entity_id: params.entityId,
        entity_kind: params.entityKind
      }
    )
    return response.data
  }

  /**
   * Продлить бафф
   */
  async extendBuff(params: {
    buffId: string
    entityId: number
    entityKind: 'user' | 'project'
    days?: number
  }): Promise<ExtendBuffResponse> {
    const response = await this.httpClient.post<ExtendBuffResponse>(
      `/buffs/${params.buffId}/extend`,
      {
        entity_id: params.entityId,
        entity_kind: params.entityKind,
        additional_days: params.days
      }
    )
    return response.data
  }

  /**
   * Откатить бафф
   */
  async revertBuff(params: {
    buffId: string
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<RevertBuffResponse> {
    const response = await this.httpClient.post<RevertBuffResponse>(
      `/buffs/${params.buffId}/revert`,
      {
        entity_id: params.entityId,
        entity_kind: params.entityKind
      }
    )
    return response.data
  }

  /**
   * Отменить бафф до применения
   */
  async cancelBuff(params: {
    buffId: string
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<CancelBuffResponse> {
    const response = await this.httpClient.post<CancelBuffResponse>(
      `/buffs/${params.buffId}/cancel`,
      {
        entity_id: params.entityId,
        entity_kind: params.entityKind
      }
    )
    return response.data
  }

  /**
   * Повторить попытку применения/отката баффа после ошибки
   */
  async retryFailedBuff(params: {
    buffId: string
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<RetryBuffResponse> {
    const response = await this.httpClient.post<RetryBuffResponse>(
      `/buffs/${params.buffId}/retry`,
      {
        entity_id: params.entityId,
        entity_kind: params.entityKind
      }
    )
    return response.data
  }

  /**
   * Вычислить эффективные лимиты
   */
  async calculateEffectiveLimits(params: {
    entityId: number
    entityKind: 'user' | 'project'
  }): Promise<EffectiveLimitsResponse> {
    const response = await this.httpClient.get<EffectiveLimitsResponse>(
      `/buffs/${params.entityKind}/${params.entityId}/effective-limits`
    )
    return response.data
  }

  /**
   * Вычислить эффективные лимиты пользователя в проекте
   */
  async getUserInProjectEffectiveLimits(params: {
    projectId: number
    userId: number
  }): Promise<EffectiveLimitsResponse> {
    const response = await this.httpClient.get<EffectiveLimitsResponse>(
      `/buffs/project/${params.projectId}/user/${params.userId}/effective-limits`
    )
    return response.data
  }

  /**
   * Получить структуру данных entity в плоском формате
   */
  async getEntityData(params: {
    entityId: number
    entityKind: 'user' | 'project'
    projectId?: number
  }): Promise<{ data_paths: Record<string, any> }> {
    const queryParams: Record<string, any> = {
      entity_id: params.entityId,
      entity_kind: params.entityKind
    }
    if (params.projectId) {
      queryParams.project_id = params.projectId
    }
    const response = await this.httpClient.get<{ data_paths: Record<string, any> }>(
      '/buffs/entity-data',
      queryParams
    )
    return response.data
  }
}

File: packages/python/src/agentstack_sdk/modules/agent_buffs.py
"""
AgentBuffs - Модуль для работы с баффами
"""

from typing import Dict, Any, Optional, List
from .agent_api import AgentAPI


class AgentBuffs(AgentAPI):
    """Модуль для работы с баффами"""
    
    async def get_active_buffs(
        self,
        entity_id: int,
        entity_kind: str = "user",
        category: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получить список активных баффов"""
        params = {"category": category} if category else {}
        return await self.http_client.get(
            f"/buffs/{entity_kind}/{entity_id}/active",
            params=params
        )
    
    async def get_buff_status(
        self,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Получить статус баффов"""
        return await self.http_client.get(
            f"/buffs/{entity_kind}/{entity_id}/status"
        )
    
    async def get_buff(
        self,
        buff_id: str,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Получить информацию о конкретном баффе"""
        return await self.http_client.get(
            f"/buffs/{buff_id}",
            params={
                "entity_id": entity_id,
                "entity_kind": entity_kind
            }
        )
    
    async def get_project_active_buffs(
        self,
        project_id: int,
        include_ecosystem: bool = True,
        category: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получить активные баффы проекта"""
        params = {
            "include_ecosystem": include_ecosystem
        }
        if category:
            params["category"] = category
        return await self.http_client.get(
            f"/buffs/project/{project_id}/active",
            params=params
        )
    
    async def get_user_in_project_active_buffs(
        self,
        project_id: int,
        user_id: int,
        include_project_buffs: bool = True,
        include_ecosystem: bool = False,
        category: Optional[str] = None
    ) -> Dict[str, Any]:
        """Получить активные баффы пользователя в проекте"""
        params = {
            "include_project_buffs": include_project_buffs,
            "include_ecosystem": include_ecosystem
        }
        if category:
            params["category"] = category
        return await self.http_client.get(
            f"/buffs/project/{project_id}/user/{user_id}/active",
            params=params
        )
    
    async def create_buff(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Создать новый бафф"""
        return await self.http_client.post("/buffs", json=data)
    
    async def apply_buff(
        self,
        buff_id: str,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Применить бафф"""
        return await self.http_client.post(
            f"/buffs/{buff_id}/apply",
            json={"entity_id": entity_id, "entity_kind": entity_kind}
        )
    
    async def extend_buff(
        self,
        buff_id: str,
        entity_id: int,
        entity_kind: str = "user",
        additional_days: Optional[int] = None
    ) -> Dict[str, Any]:
        """Продлить бафф"""
        payload = {
            "entity_id": entity_id,
            "entity_kind": entity_kind
        }
        if additional_days:
            payload["additional_days"] = additional_days
        return await self.http_client.post(
            f"/buffs/{buff_id}/extend",
            json=payload
        )
    
    async def revert_buff(
        self,
        buff_id: str,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Откатить бафф"""
        return await self.http_client.post(
            f"/buffs/{buff_id}/revert",
            json={"entity_id": entity_id, "entity_kind": entity_kind}
        )
    
    async def cancel_buff(
        self,
        buff_id: str,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Отменить бафф до применения"""
        return await self.http_client.post(
            f"/buffs/{buff_id}/cancel",
            json={"entity_id": entity_id, "entity_kind": entity_kind}
        )
    
    async def retry_failed_buff(
        self,
        buff_id: str,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Повторить попытку применения/отката баффа после ошибки"""
        return await self.http_client.post(
            f"/buffs/{buff_id}/retry",
            json={"entity_id": entity_id, "entity_kind": entity_kind}
        )
    
    async def calculate_effective_limits(
        self,
        entity_id: int,
        entity_kind: str = "user"
    ) -> Dict[str, Any]:
        """Вычислить эффективные лимиты"""
        return await self.http_client.get(
            f"/buffs/{entity_kind}/{entity_id}/effective-limits"
        )
    
    async def get_user_in_project_effective_limits(
        self,
        project_id: int,
        user_id: int
    ) -> Dict[str, Any]:
        """Вычислить эффективные лимиты пользователя в проекте"""
        return await self.http_client.get(
            f"/buffs/project/{project_id}/user/{user_id}/effective-limits"
        )
    
    async def get_entity_data(
        self,
        entity_id: int,
        entity_kind: str = "user",
        project_id: Optional[int] = None
    ) -> Dict[str, Any]:
        """Получить структуру данных entity в плоском формате"""
        params = {
            "entity_id": entity_id,
            "entity_kind": entity_kind
        }
        if project_id:
            params["project_id"] = project_id
        return await self.http_client.get(
            "/buffs/entity-data",
            params=params
        )
